### ISSUE TITLE
Add the ability to specify a different repository for the *environment* of a repo

### ISSUE THREAD

**Issue by choldgraf** [jupyter/repo2docker#917]
Over the years, we've felt a tension between *flexibility* and *speed* in Binder launches. This is most-obvious in repositories that are often updated in their *content*, but not in their *environment*. We've recommended various workarounds for this (e.g., [using nbgitpuller to separate content from environment](https://discourse.jupyter.org/t/tip-speed-up-binder-launches-by-pulling-github-content-in-a-binder-link-with-nbgitpuller/922)), but many folks spend a lot of extra time waiting for a binder session to launch just because they've changed a typo in a notebook somewhere.

I think one way that we could get around this could be to allow for users to specify an **environment repository** in their code. This could behave like this:

in `runtime.txt`:

```
environment-<URL to git repository>
```

which would trigger the following behavior:

1. All other configuration files in the current repository are ignored
2. repo2docker is called on the repo specified in the `runtime.txt` file
3. When the session begins, all of the files in the *environment repo* are removed, and replaced by the ones in the current repo

In this way, people could explicitly tag a different repository as an *environment repository* and thus save a lot of time in re-building etc. They could pin the URL of the target repository to a specific hash/branch/etc just like a normal binder repo, so best-practices in reproducibility will still function.

This could:

* Save our cloud costs, because fewer unique images would end up being built
* Save launch times, because fewer unique images == less docker pulls and repo2docker builds == less launch time
* Be a way to support a "default community image" that many people can use, which would result in *much* faster launch times (e.g., just tell people "put `environment-https://github.com/jupyterhub/community-environment` in your `runtime.txt` file)

What do people think about this?

**Comment by story645** [comment]
Could the environment repo be a submodule of the orginal repo so that I don't have to maintain two copies of environment setups? Thinking classroom case where I optimally want them to build locally of env.yaml but want to offer the binder as backup.

**Comment by betatim** [comment]
While I seriously think about this proposal, why isn't this the same as putting:
```
FROM some-base-image:124124

COPY ....
````
in the `Dockerfile` of your contents repo and building `some-base-image` from a different repo (with repo2docker).

My hunch is that those who don't care about the env don't want to have to create the env either ("fix a typo, slow rebuild" should not happen any more because of how we organise the layers). They want a ready made env that someone else maintains for them. Which makes me think we need a compacter notation for the launch URL to do this (a feature in BinderHub). Something like default.mybinder.org/v3/gh/myorg/smyrepo which is a shorthand for `https://mybinder.org/v2/gh/default-image-org/default-binder/master?urlpath=git-pull?repo=<URL_TO_MY_REPO_HERE>` (as well as variations that aren't nbgitpuller based when the content is somewhere not on github)

**Comment by danielballan** [comment]
I think this is very much needed.

To my eye, canonizing a "default" environment at the level of a subdomain (`contentonly.mybinder.org` -> `.../default-image-org/`) cuts too far against best practices [along the lines of what @choldgraf articulated on Twitter](https://twitter.com/choldgraf/status/1275865045356822528?s=20). But adding fields to mybinder.org to make it easy to join an environment repo with a separate repo of content sounds very useful.

Is it sufficient for a content-only repo to encode its associated environment in a Binder URL in, say, its README? Encoding the environment in the `runtime.txt` (as suggested in the original post) is more machine-readable, but also more "special" in a way that might be putting a toe up to the line of the rule, "The REES does not encode any Binder-specific concepts."

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/how-to-reduce-mybinder-org-repository-startup-time/4956/7

**Comment by betatim** [comment]
One thing we have learnt with repo2docker: every time we invent a special config file we regret it. This means we need to find something that is already out there (or send someone out to the world to establish our new idea as a "standard") before building it into repo2docker.

**Comment by choldgraf** [comment]
A few thoughts:

On the idea of over-riding a URL for this, instead of having a "default" environment in a sub-domain, could we special-case the "contents" part of things? e.g.:

```
https://mybinder.org/v2/gh/default-image-org/default-binder/master?content=<URL_TO_MY_REPO>
```

and then BinderHub would fetch whatever `content=` points to and over-write the base repository's contents with what is inside (or copy them w/o over-writing?)


on the idea of encoding this in a text file within the repository:

point taken that adding a new standard could potentially confuse people. Some patterns that I have noticed are in CI/CD platforms. e.g.,:

[github actions](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions) allows for things like

```
- name: My first step
  uses: docker://alpine:3.8
```

so we could use syntax like:

```
uses: <mybinder.org link>
```

in one of the configuration files (maybe `runtime.txt`?)

[CircleCI is a bit more explicit / involved](https://circleci.com/docs/2.0/custom-images/):

```
build:
    docker:
      - image: circleci/cci-demo-docker-primary:0.0.1
```

**Comment by betatim** [comment]
I think the ability to compose several images into one is out of scope for repo2docker.

There is demand for using one image with lots of different content (a la Google colab or Kaggle kernels). This is not really a task for repo2docker itself.

I am not sure there is large demand for being able to combine arbitrary "environment images" with arbitrary other "content images". I also think this is a very hard problem to solve (just thinking of all the possible cases where the two images are incompatible with each other). So I think we should leave this for someone else to solve or tools that run inside the container (like nbgitpuller).

**Comment by choldgraf** [comment]
Just to be clear, I wasn't suggesting we combine the images, I was suggesting we use the image from one, and pull in the files from the other.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/binder-config-re-design/5130/1

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/jovian-ml-increased-usage-in-binder/6215/6

**Comment by aakashns** [comment]
Please consider adding this as a built-in feature within Binder, I feel confident this can have a huge practical impact. I wish we'd found out about it sooner, and we could've avoided tends of thousands of hours of builds & people waiting on them.

A simple way to do this via the URL would be really helpful. Something along the lines of:

```
https://mybinder.org/v2/git?source_repo=xxx&env_repo=xxx
```

or 

```
https://mybinder.org/v2/git?source=example.com/url/of/notebook.ipynb&env=example.com/url/of/environment.yml
```

We'd be more than happy to work a PR for this if you'd like to pursue this. 

cc @sidujjain

**Comment by manics** [comment]
@aakashns There's scope for making nbgitpuller more easily accessible on BinderHub and mybinder.org. This is a closed issue, would you like to start a new one on https://github.com/jupyterhub/binderhub which is the relevant repo and we can discuss options?

**Comment by choldgraf** [comment]
@manics sorry about that - I encouraged @aakashns to post in here because the reason for closing this issue was "not sure there is large demand" so I thought it was relevant to that point.

**Comment by betatim** [comment]
There is a thread with a lot of context, ideas and discussion about providing a default environment (decided by the admin of the BinderHub) into which content can be pulled via nbgitpuller in https://github.com/jupyterhub/mybinder.org-deploy/issues/1474 I think that issue is the best place to read and add to.

---

If you want to use your own "base environment" already today take a look at https://github.com/betatim/kaggle-binder/tree/master#how-can-i-use-this for an example which shows how to do that (no new code needed).

A quick win regarding discoverability would be to work on https://mybinder.readthedocs.io/en/latest/howto/external_binder_setup.html which is a tutorial that suggests the nbgitpuller + env idea. Also interesting when it comes to documentation changes to improve discoverability is [an issue about merging the various Binder docs](https://github.com/jupyterhub/binderhub/issues/1091).
